cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,158 @@ def get_sbom_components(self) -> list[dict[str, str]]:
+    :raise FetchError: If download failed
+    """
+    info = extract_git_info_npm(vcs)
+    download_path = RootedPath(
+        Path(
+            download_dir,
+            info["host"],  # host
+            info["namespace"],
+            info["repo"],
+            f'{info["repo"]}-external-gitcommit-{info["ref"]}.tgz',
+        )
+    )
+
+    # Create missing directories
+    directory = os.path.dirname(download_path)
+    os.makedirs(directory, exist_ok=True)
+    clone_as_tarball(info["url"], info["ref"], download_path.path)
+
+    return download_path
+
+
+def get_dependencies(
+    download_dir: RootedPath, deps_to_download: Dict[str, Dict[str, str]]
+) -> Dict[str, RootedPath]:
+    """
+    Download npm dependencies.
+
+    Receives the destination directory (download_dir)
+    and the dependencies to be downloaded (deps_to_download).
+
+    :param download_dir: Destination directory path where deps will be downloaded
+    :param deps_to_download: Dict of dependencies to be downloaded.
+    :return: Dictionary of dependencies with downloaded paths
+    """
+    files_to_download: dict[str, dict[str, Any]] = {}
+    download_paths = {}
+    for dependency, info in deps_to_download.items():
+        if info["url"].startswith("git+"):
+            # Git source
+            download_paths[dependency] = clone_repo_pack_archive(info["url"], download_dir)
+        else:
+            if "registry.npmjs.org" == urlparse(info["url"]).hostname:
+                # Tar archive from npm registry
+                archive_name = f'{dependency}-{info["version"]}.tgz'.removeprefix("@").replace(
+                    "/", "-"
+                )
+                download_paths[dependency] = download_dir.join_within_root(archive_name)
+            else:
+                # Tar archive
+                algorithm, digest = ChecksumInfo.from_sri(info["integrity"])
+                download_paths[dependency] = RootedPath(
+                    Path(
+                        download_dir.path,
+                        f"external-{dependency}",
+                        f"{dependency}-external-{algorithm}-{digest}.tgz",
+                    )
+                )
+
+                # Create missing directories
+                directory = os.path.dirname(download_paths[dependency])
+                os.makedirs(directory, exist_ok=True)
+
+            files_to_download[info["url"]] = {
+                "download_path": download_paths[dependency],
+                "integrity": info["integrity"],
+            }
+
+    # Asynchronously download tar files
+    asyncio.run(
+        async_download_files(
+            {url: item["download_path"] for (url, item) in files_to_download.items()}, 5
+        )
+    )
+
+    # Check integrity of downloaded packages
+    for url, item in files_to_download.items():
+        must_match_any_checksum(
+            item["download_path"], [ChecksumInfo.from_sri(str(item["integrity"]))]
```

Comment:
Could we handle missing `integrity` properly?

Ideally validate before downloading

Btw, not long ago I saw a lockfile which was completely missing integrity values (in an OSBS build, this month I think)

Reply:
IIRC we decided not to fail for missing checksums, but we must still skip the verification (otherwise we will fail with an ugly error when verification the checksum "None")

Reply:
Ideally validate before downloading. Is it possible to validate checksum without having file downloaded? I skipped for now that chesum check if integrity is unavailable. Should be sbom update about missing integrity part of this PR?

Reply:
> Ideally validate before downloading

I meant "validate that integrity isn't missing", but then we decided to ignore missing integrity instead

SBOM update needs a design first (or wait for the Go checksum handling to be done, we'll probably want to report it the same way for npm)